rb_tree.py

```diff
@@ -1,81 +1,287 @@
+
+    @property
+    def is_uncolored(self) -> bool:
+        return self.color == NODE_COLOR.NIL
+
+    def __bool__(self) -> bool:
+        """ implicit version of is_colored property """
+        return self.color != NODE_COLOR.NIL
+    
+    def __repr__(self) -> str:
+        """ 
+        Spit out a few things about the node to help with
+        debugging.
+        """        
         return '{color} {val} Node'.format(color=self.color, val=self.value)
 
-    def __iter__(self):
-        if self.left.color != NIL:
+    def __str__(self) -> str:
+        """
+        For debugging purposes in depth.
+        """
+        return "id:{} L:{} <- {}:{} -> R:{} P:{}".format(
+            id(self), id(self.left), self.value, self.color, id(self.right), id(self.parent)
+            )
+
+    def iter_dump(self) -> object:
+        """
+        Just like __iter__, except that we dump out the __str__ values.
+        """
+        if self.left.color != NODE_COLOR.NIL:
+            yield from self.left.iter_dump()
+
+        yield str(self) + (" ROOT NODE " if self.parent is None else "")
+
+        if self.right.color != NODE_COLOR.NIL:
+            yield from self.right.iter_dump()
+
+    def __iter__(self) -> object:
+        """ 
+        return all the left side Nodes, then one's self,
+        then all the right side Nodes.
+        """
+
+        if self.left.color != NODE_COLOR.NIL:
             yield from self.left.__iter__()
 
         yield self.value
 
-        if self.right.color != NIL:
+        if self.right.color != NODE_COLOR.NIL:
             yield from self.right.__iter__()
 
-    def __eq__(self, other):
-        if self.color == NIL and self.color == other.color:
+    def __eq__(self, other:Node):
+        """
+        Deciding if two nodes are the same based on their
+        values and colors.
+        """
+    
+        # Case 0: The argument is not a fellow Node.
+        if not isinstance(other, Node): return NotImplemented
+
+        # Case 1: Both nodes are NIL, so we don't need to check values.
+        if self.color == NODE_COLOR.NIL and self.color == other.color:
             return True
 
+        # Case 2: Do we have the same parent? If either we or they are
+        # an orphan/root node, then we only need to check the other's 
+        # parents. Otherwise we need same parent and same value and color.
         if self.parent is None or other.parent is None:
             parents_are_same = self.parent is None and other.parent is None
         else:
-            parents_are_same = self.parent.value == other.parent.value and self.parent.color == other.parent.color
-        return self.value == other.value and self.color == other.color and parents_are_same
+            parents_are_same = ( self.parent.value == other.parent.value and 
+                        self.parent.color == other.parent.color )
+
+        return ( self.value == other.value and 
+                 self.color == other.color and 
```

Comment:
I'm not certain whether this is in the PEP8 style guide but I recall seeing it be common Python style - please place the `and` in front of the condition in all cases rather than have it be trailing